### PR TITLE
[wip][redo 9585] Support serialized checkpoint loading

### LINKED ISSUE #9406: Support serialized checkpoint loading
### Motivation

Currently, all processes load the checkpoint at the same time. This can quickly OOM for large models. These use cases, especially with things like mixture of experts, might require serialized loading of checkpoint dicts across ranks (ie load the checkpoint one rank at a time per node). Could we enable this for DDP? 

Prior work: https://github.com/PyTorchLightning/pytorch-lightning/pull/8515

### Pitch

This would be controlled per training type plugin. Example pseudocode: https://gist.github.com/ananthsub/4ceedff56b2049a63bbb05ccd283b919

To work through: 
1. Should the TrainingTypePlugin have responsibility of calling `on_load_checkpoint` instead of the Trainer/connector? This would make sense as the TTP "owns" the LightningModule inside of the trainer, and since it already offers `load_model_state_dict`: https://github.com/PyTorchLightning/pytorch-lightning/blob/41ba639859cf6c6bf319eb33e5b3394504315962/pytorch_lightning/plugins/training_type/training_type_plugin.py#L159-L160

DeepSpeed already eschews most of the checkpoint connector logic when it comes to loading the lightning module state. This could be a gap for metrics, and this means we could be calling `on_load_checkpoint` multiple times with certain plugins. In my opinion, this points to all LightningModule state load/save/alterations should sit inside of the training type plugin. 

2. Does loading one process per node run into other issues with collectives potentially being used in `LightningModule.on_load_checkpoint` ?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #9406 (redo of #9585, messed up git)

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -535,3 +537,19 @@ def teardown(self) -> None:
             self.lightning_module.cpu()
             # clean up memory
             torch.cuda.empty_cache()
+
+    def load_model_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
+        if "state_dict" not in checkpoint and self._self_deleted_checkpoint_state_dict:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp.py:542]: Why do you need `self._self_deleted_checkpoint_state_dict` ? note: The variable name isn't great :)

Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/ddp.py:542]: original PR [comment thread](https://github.com/PyTorchLightning/pytorch-lightning/pull/9585#discussion_r711380142):  I think in case `load_model_state_dict` gets called after `load_checkpoint`, @ananthsub was this the intention?

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:542]: @tchaton it's to distinguish if the 

@jjenniferdai - we should also ensure we reset the property in the `teardown` function

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -154,12 +148,8 @@ def restore_model(self) -> None:
 
     def restore_model_weights(self, checkpoint_path: Optional[_PATH]) -> None:
         """Restore only the model weights."""
-        checkpoint = self._loaded_checkpoint
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:157]: - when is checkpoint_path `None` ? 
- i haven't traced through the paths fully here. is there a concern we're now loading a different checkpoint dict somewhere?

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 3]
```diff
@@ -141,17 +141,6 @@ def wrap_policy(*args, **kwargs):
...
-    def setup_environment(self) -> None:
-        super().setup_environment()
-        model_call_configure_sharded_model_hook = getattr(
-            self.lightning_module, "call_configure_sharded_model_hook", False
-        )
-        if not model_call_configure_sharded_model_hook:
-            # if model has not called configure sharded model, we reset
-            # the training type plugin's call_configure_sharded_model_hook
-            # to give trainer a chance to configure.
-            self.call_configure_sharded_model_hook = True
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:153]: on rebase this should be gone now

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 4]
```diff
@@ -153,7 +153,10 @@ def results(self) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
 
     def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
         torch.cuda.empty_cache()
-        return self.checkpoint_io.load_checkpoint(checkpoint_path)
+        checkpoint = self.checkpoint_io.load_checkpoint(checkpoint_path)
+        self.lightning_module.on_load_checkpoint(checkpoint)
+        self.load_model_state_dict(checkpoint)
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:158]: n00b question: do you think this ought to be on the base class? vs calling `self.lightning_module.on_load_checkpoint(checkpoint)` in `load_model_state_dict` ?

Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:158]: do you mean if `on_load_checkpoint` should be in `load_model_state_dict` v.s. `load_checkpoint` here in ttp base class? if so I was thinking to maybe keep `load_model_state_dict` only performing `load_state_dict` as is, wdyt?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 2 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:157]: its never None, `restore_model_weights` is only called with `trainer._ckpt_path` if `trainer._ckpt_path` is not None. should we remove the None check for clarity?

Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:157]: in that case i believe we should. @awaelchli what do you think?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:158]: `load_checkpoint` is materializing the file contents in memory whereas `load_model_state_dict` is specifically on reloading the model's state

in my head, i think `on_load_checkpoint` is in spirit closer to `load_model_state_dict` than it is to loading the file...but i'm not 100% certain if my assumption is right. that's based on the current codebase where `model.on_load_checkpoint` is called right before the state dict.

if this was redone, maybe the name for this hook would've been `on_load_state_dict` ..

@tchaton @awaelchli does my understanding sqaure up with yours?

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 1 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/ddp.py:544]: "but later, the checkpoint connector will call `load_model_state_dict`" - this doesn't happen anymore after these checkpoint connector changes right?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:177]: hm agree `on_load_checkpoint` is closer to `load_model_state_dict` than `load_checkpoint (file)` - maybe some renaming?
- old `load_checkpoint` --> `load_checkpoint_file`
- new method `load_checkpoint_file_and_model_state` (a better name than this?) = `load_checkpoint_file`, `on_load_checkpoint`, `load_model_state_dict`

Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:177]: just to confirm - like the below? (so `load_model_state_dict`, `load_checkpoint` will have different functionalities across different TTPs)

base TTP:
- `load_model_state_dict` = `on_load_checkpoint`, `model.load_state_dict`
- `load_checkpoint` = load checkpoint file

serialized loading TTPs:
- `load_model_state_dict` = `model.load_state_dict` only
- `load_checkpoint` = for rank in ranks, `super().load_checkpoint`, `on_load_checkpoint`, `load_model_state_dict`
- (`load_model_state_dict` will be called again in checkpoint connector, but early exits with the `_self_deleted_checkpoint_state_dict` boolean)

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:177]: yes that's correct, at least this is my preference. what do you think @jjenniferdai @awaelchli @carmocca ?

Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:177]: I slightly prefer the second option because it's more consistent across TTPs. but defer to others more experienced :)

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:544]: Sounds good to me, but let's find a better variable name. Something like `self._has_loaded_state_dict`

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:177]: was pinged here but sorry, I'm not able to follow the conversation.
- you can rename things in the io plugin because its not released yet
- we are not allowed to change the order of hook calls

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -984,9 +984,6 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
         self.data_connector.prepare_data()
         self.callback_connector._attach_model_callbacks()
 
-        if self._ckpt_path and not self.accelerator.restore_checkpoint_after_pre_dispatch:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:987]: this was here for a reason. check deepspeed vs. training type base:

deepspeed: restore_checkpoint_after_pre_dispatch = True
all others: restore_checkpoint_after_pre_dispatch = False

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:177]: no change to order of hook calls or io plugin, but moving calling responsibility of `LightningModule.on_load_checkpoint` from checkpoint connector to TTP ("to work through" portion of issue #9406)

2 different options in ([comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/9605#discussion_r717806551)) on how to organize this:
- first option = this [comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/9605#discussion_r717165512)
- second option = this PR currently

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/trainer.py:987]: hm requiring validate/test/predict to restore model weights before setup while fit restores after seems like an odd reliance, does anyone have context/insights on use cases that would rely on this gap? originally introduced in PR #8352
